Migrate/group member list

Pasindufdo98 opened the pull request:
# Migration : Group Member List

# Description

This component displays the list of members in a selected group, allowing users to add new members or remove existing ones. Also manages related actions such as showing member details and updating the list when group data changes. 

Before:
<img width="1627" height="961" alt="image" src="https://github.com/user-attachments/assets/c3417ff0-7e22-4539-b0b2-7903e8260c10" />

After:
<img width="1918" height="1137" alt="image" src="https://github.com/user-attachments/assets/934d58e7-0e14-413c-a8fa-d100d129c0b1" />


Fixes # (issue)

# How Has This Been Tested?

Tested by adding and removing members across different groups to confirm the list updates correctly and toolbar actions respond based on user role. 

## Testing Checklist:

- [x] Tested in latest Chrome
- [ ] Tested in latest Safari
- [x] Tested in latest Firefox

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have requested a review on the Pull Request

Pasindufdo98 committed:
NEW: Create initial files for migration of group-member-list

Pasindufdo98 committed:
refactor: downgrade angular member-list component

Pasindufdo98 committed:
NEW: implemented angular group-member-list

Pasindufdo98 committed:
NEW: added scss content

Pasindufdo98 committed:
Restore edit.tpl.html (exclude unintended change)

Pasindufdo98 committed:
Removed unlinked old angularjs files

returnMarcco commented on src/app/doubtfire-angularjs.module.ts:
In this file, you'll need to remove the `import` statement for the old implementation of the `group-member-list.js` file. I'll comment this out for now on my end so I can continue with the review.

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.ts:
This class looks good overall. I'd suggest adding some proper typing to class properties such as the `@Inputs` and `@Outputs`.

returnMarcco reviewed (changes requested):
Hi @Pasindufdo98,

This is a peer review (junior OR senior) for these changes.

All in all, the component is functioning as expected. I am able to `add` and `remove` members to and from a `group`, respectively. Good work. Below are some things that will need changing/investigating:

1. Please remove the reference to the old `.js` component-file in `doubtfire-angularjs.module.ts`. This is causing the app to not load.

2. I've noticed what seems to be potentially infinite triggering of the `members()` getter method highlighted in the below screenshot. I'm not sure if this behavior was present in the old component, but you can reproduce it by creating a `group`, and by putting a breakpoint on the method name shown in the below screenshot, this method should trigger on its own. If you step through the code in the debugger, you should notice that this method is constantly called in succession:
<img width="1891" height="943" alt="Screenshot 2025-08-17 151204" src="https://github.com/user-attachments/assets/ff47cfe7-f19e-4534-a3e9-4e182969b433" />

3. The `-` sign on the `Remove` button seems like it needs a space between itself and the `Remove` text.
<img width="971" height="132" alt="Screenshot 2025-08-17 150658" src="https://github.com/user-attachments/assets/5c02b80d-94ce-4683-9c75-eae51d60bed9" />

Once these requested changes have been made/investigated, please leave your response in this PR and re-request a review from me from the PR `Reviewers` list.


Cheers

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.html:
```diff
@@ -0,0 +1,81 @@
+        <button class="btn btn-sm btn-danger" (click)="removeMember(m)">
+          <i class="fa fa-minus"></i>
+          <span *ngIf="project && m?.id === project?.id">Leave</span>
+          <span *ngIf="!project || m?.id !== project?.id">Remove</span>
```
This `Remove` button seems to require `whitespace` between the `-` icon and the `Remove` text.

Pasindufdo98 replied:
Hey Jason, I fixed the spacing issue. Thanks for pointing out.

returnMarcco commented:
Hi @Pasindufdo98,

Can you please update the `Migration Progress List` section of README.MD.

Cheers

Pasindufdo98 committed:
Fixed repeated getter calls in group-member-list by replacing template getters with fields, enabling OnPush change detection, and making member removal update the UI instantly.

Pasindufdo98 commented:
> Hi @Pasindufdo98,
> 
> This is a peer review (junior OR senior) for these changes.
> 
> All in all, the component is functioning as expected. I am able to `add` and `remove` members to and from a `group`, respectively. Good work. Below are some things that will need changing/investigating:
> 
> 1. Please remove the reference to the old `.js` component-file in `doubtfire-angularjs.module.ts`. This is causing the app to not load.
> 2. I've noticed what seems to be potentially infinite triggering of the `members()` getter method highlighted in the below screenshot. I'm not sure if this behavior was present in the old component, but you can reproduce it by creating a `group`, and by putting a breakpoint on the method name shown in the below screenshot, this method should trigger on its own. If you step through the code in the debugger, you should notice that this method is constantly called in succession:
> 
> <img alt="Screenshot 2025-08-17 151204" width="1891" height="943" src="https://private-user-images.githubusercontent.com/106223269/478743212-ff47cfe7-f19e-4534-a3e9-4e182969b433.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYwMzI5NTgsIm5iZiI6MTc1NjAzMjY1OCwicGF0aCI6Ii8xMDYyMjMyNjkvNDc4NzQzMjEyLWZmNDdjZmU3LWYxOWUtNDUzNC1hM2U5LTRlMTgyOTY5YjQzMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNFQxMDUwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNjdjYWM4Njk4NDMzNmZhYTM2NjM2ZGM3NWU4NTI0NTBiOGM5MGJiMzA3Mzg2YzUyMGQ4NzE2N2MyYzczNzgwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bEE45wNMGyng_Wo31sfZEFUr3w2ZR-7XBSQleceb_CI">
> 3. The `-` sign on the `Remove` button seems like it needs a space between itself and the `Remove` text.
> 
> <img alt="Screenshot 2025-08-17 150658" width="971" height="132" src="https://private-user-images.githubusercontent.com/106223269/478743238-5c02b80d-94ce-4683-9c75-eae51d60bed9.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYwMzI5NTgsIm5iZiI6MTc1NjAzMjY1OCwicGF0aCI6Ii8xMDYyMjMyNjkvNDc4NzQzMjM4LTVjMDJiODBkLTk0Y2UtNDY4My05Yzc1LWVhZTUxZDYwYmVkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNFQxMDUwNThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MzhiMWJjYzNmMWI1MjVlNWUzMzcxMDRkNDM1NGM4ZTQxMmE4MTM4YjRhY2UwNzA3Yzg5MDk4OTlkM2U3YjU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.XED64cFUFbiokVrzNqYcS6081_7XrA4MuDncCaJeM0k">
> Once these requested changes have been made/investigated, please leave your response in this PR and re-request a review from me from the PR `Reviewers` list.
> 
> Cheers

Hey @returnMarcco ,
Thanks for the catch Jason, the repeated calls were from the template getter being re-evaluated every change-detection pass. I did the fixes as follows and added some more improvements too.

### Fixes implemented

- Replaced get members()/sortedMembers() with fields, recalculated via resort() only when data or sort order changes.
- Enabled ChangeDetectionStrategy.OnPush and added markForCheck() after async loads to trigger view updates when needed.
- Updated removeMember to refresh local arrays so rows disappear right away, no page refresh needed.
- Verified (members-loaded) output correctly toggles the toolbar.
- The component now updates only when expected and avoids unnecessary cycles too.

<img width="1918" height="1023" alt="image" src="https://github.com/user-attachments/assets/cbb2ee13-322b-42a2-8e9f-b9dab66319ed" />

disururathnayake commented:
Hi @Pasindufdo98 ,

I have successfully pulled and reviewed your code. Nice work on the migration. The component is working as expected

returnMarcco reviewed (commented):
Hi @Pasindufdo98,

When I attempt to add a `student` to a `group`, that student doesn't seem to show in the list for me. Can you please check to see if you're experiencing the same?

Also, when I add a `group` to a `group set` and reload the page, the group  no longer shows in the list (it exists in the database though). Can you confirm if this is the case for you? This probably isn't a result of your changes though.

Cheers

Pasindufdo98 committed:
fix(group-member-list): auto-detect member changes

Pasindufdo98 commented:
> Hi @Pasindufdo98,
> 
> When I attempt to add a `student` to a `group`, that student doesn't seem to show in the list for me. Can you please check to see if you're experiencing the same?
> 
> Also, when I add a `group` to a `group set` and reload the page, the group no longer shows in the list (it exists in the database though). Can you confirm if this is the case for you? This probably isn't a result of your changes though.
> 
> Cheers

Hi @returnMarcco ,
Thanks for reviewing again.
- I fixed the issue with not updating member list in the frontend. Now it should be working when adding and removing a member without reloading.
- When adding a `group` to a `group set`, it shows immediately without any error for me. 
<img width="1920" height="1116" alt="Screenshot (1715)" src="https://github.com/user-attachments/assets/d8778be3-3637-44ae-9a9a-e8bf1ee9f06f" />

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.ts:
```diff
@@ -1,36 +1,63 @@
     }
   }
 
+  private resort(): void {
```
Hi @Pasindufdo98,

Can you please annotate this method with a JSDoc and explain what it does as it's logic is fairly complex. Also, I strongly suggest replacing vars/params like `a` and `k` and `vb` etc. with more descriptive names.

Cheers

Pasindufdo98 replied:
Hi @returnMarcco , 
I have added an annotation with a JSDoc. Thanks for the pointing out that.

Pasindufdo98 committed:
fix(group-member-list): show add/remove changes without refresh

Pasindufdo98 commented:
> Hi @Pasindufdo98,
> 
> When I attempt to add a `student` to a `group`, that student doesn't seem to show in the list for me. Can you please check to see if you're experiencing the same?
> 
> Also, when I add a `group` to a `group set` and reload the page, the group no longer shows in the list (it exists in the database though). Can you confirm if this is the case for you? This probably isn't a result of your changes though.
> 
> Cheers

Hey @returnMarcco,
I noticed this issue only occurs when signing in as an admin. I also tested it on the old AngularJS version and the same problem exists there, so it’s not introduced by my migration changes to group-member-list. When logged in as a convenor, the functionality seems to work as expected.

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.html:
Because of recent feedback from `doubtfire-lms`, I suggest the following:
1. Replacing components here with `Angular Material` equivalents: https://material.angular.dev/components/categories
    - I would also consider swapping out the `Table` element with the `Angular Material` equivalent.
2. Replacing font-awesome (fa) icons with `mat-icons`.

Pasindufdo98 replied:
Hi @returnMarcco , 
I have replaced them with Angular Material. Thank you

<img width="1917" height="1132" alt="image" src="https://github.com/user-attachments/assets/4074a547-63b5-46f0-b449-58f1e1ace72d" />

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.ts:
```diff
@@ -0,0 +1,156 @@
+  @Output() membersLoaded = new EventEmitter<void>();
+  @Output() unitRoleChange = new EventEmitter<UnitRole>();
+
+
```
Please lint your code by running `Prettier`. This will fix the whitespaces and other formatting inconsistencies.

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.ts:
```diff
@@ -0,0 +1,156 @@
+        this.maybeClearRemoved(g)
+        this.changeDetectorRef.markForCheck();
+      },
+      error: () => {
```
Consider adding an error alert using the `alertService.alert()` method like in the original Coffeescript/AngularJS implementation.

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.ts:
```diff
@@ -0,0 +1,156 @@
+    });
+  }
+
+  private updateCanRemoveMembers(): void {
```
I am still able to remove members from a group even though that group is `locked`. Please review this functionality.

Pasindufdo98 replied:
Hey @returnMarcco , 
I tested this behaviour in the original AngularJS/coffeescript version. That behaviour existed in that too. Although if a group is set as locked, it doesn't allow to add a member but, allows to remove a member.
<img width="1920" height="1122" alt="image" src="https://github.com/user-attachments/assets/687aeeba-e832-48d1-9522-e8466a757b9a" />

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.html:
```diff
@@ -0,0 +1,81 @@
+        </button>
+      </th>
+
+      <th *ngIf="canRemoveMembers">Actions</th>
```
This is showing even if the group is locked.

Pasindufdo98 replied:
Hi @returnMarcco 
I’ve updated the condition, so the Actions column now enables when the group is unlocked. Thanks for pointing this out.

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.ts:
```diff
@@ -0,0 +1,156 @@
+  private fetchMembers(): void {
+    this.loaded = false;
+    const g = this.selectedGroup!;
+    // if (this.selectedGroup?.projectsCache?.clear) {
```
Please remove the commented out code if its not needed.

returnMarcco reviewed (changes requested):
Hey Pasindu,

I've left a few comments as feedback. Reach out if you have any questions.

Cheers

Pasindufdo98 committed:
group-member removing UI bug fix

Pasindufdo98 committed:
refactor: improve member sorting and cleanup logic

Pasindufdo98 committed:
refactor: replaced Bootstrap UI with Angular Material and fixed linting with Prettier

Pasindufdo98 committed:
add alert service for error handling and update table headers styling

Pasindufdo98 committed:
disable action column when a group is locked

Pasindufdo98 committed:
disable action column when a group is locked-updated html

returnMarcco commented on src/app/groups/group-member-list/group-member-list.component.html:
```diff
@@ -0,0 +1,77 @@
+  <ng-container matColumnDef="actions" *ngIf="displayedColumns.includes('actions')">
+    <th mat-header-cell *matHeaderCellDef>Actions</th>
+    <td mat-cell *matCellDef="let m">
+      <button mat-raised-button color="warn" (click)="removeMember(m)" [disabled]="selectedGroup?.locked">
```
Good job on the Material UI refactor, it looks good. As a last change, I'd suggest changing the icon to a `bin` like the other `Delete/Remove` buttons in the app. I'd also style the text `white`.

Pasindufdo98 replied:
Updated.
<img width="1582" height="862" alt="image" src="https://github.com/user-attachments/assets/000616ff-f6b5-4147-8998-609297aede1f" />

returnMarcco reviewed (approved):
Pre-approving this PR. Please see last recommended change. After making this change, please contact Brian as soon as possible so he can review this before opening an upstream PR. Good work.

Pasindufdo98 committed:
changed the remove btn icon and the text to white

Pasindufdo98 committed:
Merge branch '9.x' into migrate/group-member-list

Pasindufdo98 commented:
Latest Appearance
<img width="1582" height="862" alt="image" src="https://github.com/user-attachments/assets/b618f99e-aab2-4432-a2bf-6ddfa20970be" />

chelaz1234 reviewed (approved):
Hi @Pasindufdo98,

Nice, thorough migration of group-member-list with clear commit structure. All functions are work as expected. Overall good job!

disururathnayake reviewed (approved):
Hi @Pasindufdo98 ,

I have successfully pulled and reviewed your code. Nice work on the migration. The component is working as expected
